blockdag/reachability.go

```diff
@@ -547,9 +811,276 @@ func (dag *BlockDAG) updateReachability(node *blockNode, selectedParentAnticone
+
+	if !reindexRoot.isAncestorOf(newTreeNode) {
+		commonAncestor := reindexRoot.findCommonAncestor(newTreeNode)
+		return commonAncestor, nil, true, nil
+	}
+
+	chosenReindexRootChild, err := reindexRoot.findReachabilityTreeAncestorInChildren(newTreeNode)
+	if err != nil {
+		return nil, nil, false, err
+	}
+	if newTreeNode.blockNode.blueScore-chosenReindexRootChild.blockNode.blueScore < reachabilityReindexWindow {
+		return nil, nil, false, nil
+	}
+	modifiedTreeNodes, err = dag.concentrateReachabilityTreeIntervalAroundReindexRootChild(reindexRoot, chosenReindexRootChild)
+	if err != nil {
+		return nil, nil, false, err
+	}
+
+	return chosenReindexRootChild, modifiedTreeNodes, true, nil
+}
+
+// findReachabilityTreeAncestorInChildren finds the reachability tree child
+// of rtn that is the ancestor of node.
+func (rtn *reachabilityTreeNode) findReachabilityTreeAncestorInChildren(node *reachabilityTreeNode) (*reachabilityTreeNode, error) {
+
+	rootChildrenFutureCoveringSet := futureCoveringBlockSetFromReachabilityTreeNodes(rtn.children)
+	i := rootChildrenFutureCoveringSet.findIndex(&futureCoveringBlock{blockNode: node.blockNode, treeNode: node})
+	if i == 0 {
+		return nil, errors.Errorf("rtn is not an ancestor of node")
+	}
+
+	return rootChildrenFutureCoveringSet[i-1].treeNode, nil
+}
+
+func (dag *BlockDAG) concentrateReachabilityTreeIntervalAroundReindexRootChild(
+	reindexRoot *reachabilityTreeNode, chosenReindexRootChild *reachabilityTreeNode) (
+	map[*reachabilityTreeNode]struct{}, error) {
+
+	modifiedTreeNodes := make(map[*reachabilityTreeNode]struct{})
+
+	reindexRootChildNodesBeforeChosen, reindexRootChildNodesAfterChosen, err :=
+		reindexRoot.splitChildrenAroundChosenChild(chosenReindexRootChild)
+	if err != nil {
+		return nil, err
+	}
+
+	reindexRootChildNodesBeforeChosenSizesSum, modifiedNodes, err :=
+		dag.tightenReachabilityTreeIntervalsBeforeChosenReindexRootChild(reindexRoot, reindexRootChildNodesBeforeChosen)
+	if err != nil {
+		return nil, err
+	}
+	for node := range modifiedNodes {
+		modifiedTreeNodes[node] = struct{}{}
+	}
+
+	reindexRootChildNodesAfterChosenSizesSum, modifiedNodes, err :=
+		dag.tightenReachabilityTreeIntervalsAfterChosenReindexRootChild(reindexRoot, reindexRootChildNodesAfterChosen)
+	if err != nil {
+		return nil, err
+	}
+	for node := range modifiedNodes {
+		modifiedTreeNodes[node] = struct{}{}
+	}
+
+	modifiedNodes, err = dag.expandReachabilityTreeIntervalInChosenReindexRootChild(
+		reindexRoot, chosenReindexRootChild, reindexRootChildNodesBeforeChosenSizesSum, reindexRootChildNodesAfterChosenSizesSum)
+	if err != nil {
+		return nil, err
+	}
+	for node := range modifiedNodes {
+		modifiedTreeNodes[node] = struct{}{}
+	}
+
+	return modifiedTreeNodes, nil
+}
+
+func (rtn *reachabilityTreeNode) splitChildrenAroundChosenChild(chosenChild *reachabilityTreeNode) (
+	nodesBeforeChosen []*reachabilityTreeNode, nodesAfterChosen []*reachabilityTreeNode, err error) {
+
+	chosenIndex := -1
```

Comment:
findIndex logic from futureCoveringSet can be used here as well to find the index of chosen in log time (although this is probably in overkill because children are accessed linearly in the calling method anyway)

Reply:
I think I prefer keeping it the way it is, since findIndex searches for ancestors rather than direct parent-child relationships.